x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,350 @@
+                            long newFrom = lastOp.seqNo() + 1;
+                            long newSize = targetSeqNo - lastOp.seqNo();
+                            LOGGER.trace("{} received [{}] as last op while [{}] was expected, continue to read [{}/{}]...",
+                                params.getFollowShardId(), lastOp.seqNo(), targetSeqNo, newFrom, size);
+                            sendShardChangesRequest(newFrom, newSize, targetSeqNo);
+                        } else {
+                            numConcurrentReads--;
+                        }
+                    } else {
+                        lastRequestedSeqno = Math.max(lastRequestedSeqno, lastOp.seqNo());
+                        LOGGER.trace("{} post updating lastRequestedSeqno to [{}]", params.getFollowShardId(), lastRequestedSeqno);
+                        numConcurrentReads--;
+                    }
+                    if (numConcurrentWrites == 0) {
+                        coordinateWrites();
+                    }
+                    coordinateReads();
+                }
+                assert numConcurrentReads >= 0;
+            }
+        });
+    }
+
+    private void sendBulkShardOperationsRequest(Translog.Operation[] operations) {
+        innerSendBulkShardOperationsRequest(operations,
+            followerLocalCheckpoint -> {
+                retryCounter.set(0);
+                handleResponse(followerLocalCheckpoint);
+            },
+            e -> handleFailure(e, () -> sendBulkShardOperationsRequest(operations))
+        );
+    }
+
+    private synchronized void handleResponse(long followerLocalCheckpoint) {
+        processedGlobalCheckpoint = Math.max(processedGlobalCheckpoint, followerLocalCheckpoint);
+        numConcurrentWrites--;
+        assert numConcurrentWrites >= 0;
+        coordinateWrites();
+    }
+
+    private synchronized void maybeUpdateMapping(Long minimumRequiredIndexMetadataVersion, Runnable task) {
+        if (currentIndexMetadataVersion >= minimumRequiredIndexMetadataVersion) {
+            LOGGER.trace("{} index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            task.run();
+        } else {
+            LOGGER.debug("{} updating mapping, index metadata version [{}] is lower than minimum required index metadata version [{}]",
+                params.getFollowShardId(), currentIndexMetadataVersion, minimumRequiredIndexMetadataVersion);
+            updateMapping(imdVersion -> {
+                retryCounter.set(0);
+                currentIndexMetadataVersion = imdVersion;
+                task.run();
+            });
+        }
+    }
+
+    void handleFailure(Exception e, Runnable task) {
+        assert e != null;
+        if (shouldRetry(e)) {
+            if (isStopped() == false && retryCounter.incrementAndGet() <= RETRY_LIMIT) {
```

Comment:
+1 to the statement and other problems previously mentioned (a network hick up will cause a failure for each outstanding request incrementing the counter by more than it can take). I asked Martijn to think about the retry counter as a follow up in order to reduce the scope of this PR.

Reply:
I will think about this and address this in a followup pr.